### PR TITLE
Fixed avro schema decode error in functions

### LINKED ISSUE #5503: [Schema]GenericData$Record cannot be cast
**Describe the bug**

**To Reproduce**
```
@Builder
    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Foo2 {
        private String name;
        private Integer id;
    }
PulsarClient client = PulsarClient.builder()
                        .serviceUrl("pulsar://localhost:6650")
                        .build();

            AvroSchema<Foo2> schema = AvroSchema.of(SchemaDefinition.<Foo2>builder().withJsonDef("{\n" +
                    "    \"type\": \"record\",\n" +
                    "    \"name\": \"Test\",\n" +
                    "    \"fields\": [\n" +
                    "      {\n" +
                    "        \"name\": \"id\",\n" +
                    "        \"type\": [\n" +
                    "          \"null\",\n" +
                    "          \"int\"\n" +
                    "        ]\n" +
                    "      },\n" +
                    "      {\n" +
                    "        \"name\": \"name\",\n" +
                    "        \"type\": [\n" +
                    "          \"null\",\n" +
                    "          \"string\"\n" +
                    "        ]\n" +
                    "      }\n" +
                    "    ]\n" +
                    "  }").withPojo(Foo2.class).build());
            Consumer<Foo2> consumer = client.newConsumer(schema)
                .topic("pulsar-mysql-jdbc-sink-topic")
                .subscriptionName("my-subscription-2")
                .subscribe();

            while (true) {
            // Wait for a message
            Message<Foo2> msg = consumer.receive();
                Foo2 f = msg.getValue();
            }
```

2. See error

```
Exception in thread "main" java.lang.ClassCastException: org.apache.pulsar.shade.org.apache.avro.generic.GenericData$Record cannot be cast to io.streamnative.KeyValueSchemaTest$Foo2
	at io.streamnative.KeyValueSchemaTest.testConsumerByPythonProduce(KeyValueSchemaTest.java:412)
	at io.streamnative.KeyValueSchemaTest.main(KeyValueSchemaTest.java:305)
```

**Expected behavior**

**Desktop (please complete the following information):**
 - OS: [e.g.Mac]

**Additional context**

### PR DESCRIPTION
**Pull request by gaoran10**
Fixes #5503 

### Motivation

In functions, it will encounter ```ClassCastException``` when using the Avro schema for topics. 
```
Exception in thread "main" java.lang.ClassCastException: org.apache.pulsar.shade.org.apache.avro.generic.GenericData$Record cannot be cast to io.streamnative.KeyValueSchemaTest$Foo2
	at io.streamnative.KeyValueSchemaTest.testConsumerByPythonProduce(KeyValueSchemaTest.java:412)
	at io.streamnative.KeyValueSchemaTest.main(KeyValueSchemaTest.java:305)
```

### Modifications

In functions, when using Avro schema specific the ```ClassLoader``` for ```ReflectDatumReader```.

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AvroSchema.java [hunk 1]
```diff
@@ -88,14 +97,19 @@ public boolean supportSchemaVersioning() {
         return new AvroSchema<>(parseSchemaInfo(schemaDefinition, SchemaType.AVRO));
     }
 
+    public static <T> AvroSchema<T> ofWithClass(SchemaDefinition<T> schemaDefinition) {
```
Comment (sijie) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AvroSchema.java]: Do we need a separate static method? Why can we just check if schemaDefinition has POJO? If it has POJO, when we create a schema instance, we should include the POJO.

Comment (gaoran10) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AvroSchema.java]: Yes, it's more reasonable.

**Comment by sijie** [comment]
@gaoran10 can you check the integration tests?

**Comment by tuteng** [comment]
@gaoran10 Please try to rebase master for pass the test

**Comment by tuteng** [comment]
@gaoran10  You can try command: 

```
mvn clean install -DskipTests
mvn -B -f tests/docker-images/pom.xml install -am -Pdocker -DskipTests
mvn -B -f tests/pom.xml test -DintegrationTestSuiteFile=pulsar-process.xml -DintegrationTests -DredirectTestOutputToFile=false -Dgroups=function
```

**Comment by gaoran10** [comment]
> @gaoran10 can you check the integration tests?

ok, i will fix it.

**Comment by gaoran10** [comment]
@sijie Ok, I will fix it.

**Comment by gaoran10** [comment]
@tuteng Thanks, I will fix it.

**Comment by gaoran10** [comment]
This fork repository is deleted by mistake, so make a new fork and take this PR  to #6662.

**Pull request closed without merging**